ndcube/meta.py

```diff
@@ -0,0 +1,466 @@
+                        f"or the same shape as associated data axes {tuple(self.data_shape[axis])}. "
+                        f"val shape = {val.shape if hasattr(val, 'shape') else (len(val),)}\n"
+                        "We recommend using the 'add' method to set values.")
+        super().__setitem__(key, val)
+
+    @property
+    def slice(self):
+        # Docstring in ABC.
+        return _NDMetaSlicer(self)
+
+    def rebin(self, bin_shape):
+        """
+        Adjusts axis-aware metadata to stay consistent with a rebinned `~ndcube.NDCube`.
+
+        This is done by simply removing the axis-awareness of metadata associated with
+        rebinned axes. The metadata itself is not changed or removed. This operation
+        does not remove axis-awareness from metadata only associated with non-rebinned
+        axes, i.e. axes whose corresponding entries in ``bin_shape`` are 1.
+
+        Parameters
+        ----------
+        bin_shape : array-like
+            The number of pixels in a bin in each dimension.
+        """
+        # Sanitize input
+        bin_shape = np.round(bin_shape).astype(int)
+        data_shape = self.data_shape
+        bin_shape = bin_shape[:len(data_shape)]  # Drop info on axes not defined by NDMeta.
+        if (np.mod(data_shape, bin_shape) != 0).any():
+            raise ValueError("bin_shape must be integer factors of their associated axes.")
+        # Remove axis-awareness from grid-aligned metadata associated with rebinned axes.
+        rebinned_axes = set(np.where(bin_shape != 1)[0])
+        new_meta = copy.deepcopy(self)
+        null_set = set()
+        for name, axes in self.axes.items():
+            if (_is_grid_aligned(self[name], data_shape[axes])
+                and set(axes).intersection(rebinned_axes) != null_set):
+                del new_meta._axes[name]
+        # Update data shape.
+        new_meta._data_shape = new_meta._data_shape // bin_shape
+        return new_meta
+
+
+class _NDMetaSlicer:
+    """
+    Helper class to slice an NDMeta instance using a slicing item.
+
+    Parameters
+    ----------
+    meta: `NDMetaABC`
+        The metadata object to slice.
+    """
+    def __init__(self, meta):
+        self.meta = meta
+
+    def __getitem__(self, item):
+        data_shape = self.meta.data_shape
+        if len(data_shape) == 0:
+            raise TypeError("NDMeta object does not have a shape and so cannot be sliced.")
+
+        new_meta = copy.deepcopy(self.meta)
+        naxes = len(data_shape)
+        if isinstance(item, (numbers.Integral, slice)):
+            item = [item]
+        if len(item) < naxes:
+            item = np.array(list(item) + [slice(None)] * (naxes - len(item)), dtype=object)
+        elif len(item) > naxes:
+            # If item applies to more axes than have been defined in NDMeta,
+            # ignore items applying to those additional axes.
+            item = np.array(item[:naxes])
```

Comment:
This could happen now because `data_shape` is determined only from the metadata in `NDMeta` unless explicitly set by the user.  So imagine the user does not explicitly give the `data_shape`, and the `NDMeta` is applied to 4-D, but there is only metadata that applies to the 2nd axis.  In that case, the `data_shape` will be `(0, n)` where `n` is the length of the 2nd axis.  But a valid `item` could be a `tuple` of 4 `int`/`slice`, because the data has 4 axes.  To make things simple, we want the `item` to be the same length as the `data_shape`.  So we discard the trailing elements `item` as they will not be used.

Reply:
Alternatively, we could allow slicing to adjust `data_shape`.  So if `data_shape` starts as `(0, 10)`, a slice item of `(slice(0, 3), slice(0, 6), slice(None), slice(0, 4))` would lead to a new data shape of `(3, 6, 0, 4)`.

Reply:
@Cadair Are we happy with this as is?

Reply:
yeah makes sense.